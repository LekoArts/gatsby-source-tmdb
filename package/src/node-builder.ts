import { Options } from "got"
import * as TMDBPlugin from "./types/tmdb-plugin"
import * as Response from "./types/response"
import { getParam, modifyURL } from "./api-utils"
import { ERROR_CODES, IMAGE_TYPES } from "./constants"

export const imageTransformation = ({ node, configuration }: TMDBPlugin.ImageTransformation) => {
  const baseUrl = configuration.images.secure_base_url
  const imageSizes = configuration.images
  const modifiedNode = node

  // For each imageType, e.g. "backdrop_path" extend the string to an object
  // With the "source" as the original path and then all available sizes as new keys
  IMAGE_TYPES.forEach((type) => {
    if (node[`${type}_path`]) {
      modifiedNode[`${type}_path`] = imageSizes[`${type}_sizes`].reduce(
        (o, key) => ({ ...o, [key]: `${baseUrl}${key}${node[`${type}_path`]}` }),
        { source: node[`${type}_path`] as string }
      )
    }
    // When a list is queried it has all its items in "items"
    // These nodes also need to be adjusted
    if (node.items) {
      node.items.forEach((item, index) => {
        if (item[`${type}_path`]) {
          modifiedNode.items[index][`${type}_path`] = imageSizes[`${type}_sizes`].reduce(
            (o, key) => ({ ...o, [key]: `${baseUrl}${key}${item[`${type}_path`]}` }),
            { source: item[`${type}_path`] as string }
          )
        }
      })
    }
  })

  return modifiedNode
}

export const nodeBuilder = async ({
  endpoint,
  tmdbGot,
  nodeHelpers,
  pluginOptions,
  accountId,
  gatsbyApi,
  configuration,
}: TMDBPlugin.NodeBuilder) => {
  // The account_id shouldn't be in the typeName
  const urlWithoutAccountId = endpoint.url.replace(`/:account_id`, ``)
  // endpoint.context is a free-form field that people can pass in, e.g. for tv/:tv_id/season/:season_number/episode/:episode_number
  // They can pass in tv_id etc.
  const typeName = endpoint.typeName || modifyURL(urlWithoutAccountId, endpoint.context)
  const Node = nodeHelpers.createNodeFactory(typeName)

  const itemTimer = gatsbyApi.reporter.activityTimer(`Source from TMDB API for ${typeName}`)
  itemTimer.start()

  const defaults: Options = {
    searchParams: {
      language: pluginOptions.language,
      region: pluginOptions.region,
      timezone: pluginOptions.timezone,
      page: 1,
      ...endpoint.searchParams,
    },
    context: {
      account_id: accountId,
      ...endpoint.context,
    },
    pagination: {
      countLimit: endpoint.countLimit || 60,
    },
  }

  let items: Response.PaginationItems = []

  const fetchPaginatedData = async (): Promise<Response.PaginationItems> =>
    tmdbGot.paginate.all(endpoint.url, {
      responseType: `json`,
      context: defaults.context,
      searchParams: defaults.searchParams,
      pagination: {
        countLimit: defaults.pagination.countLimit,
        // @ts-ignore
        transform: (response) => {
          const { results } = response.body as Response.PaginationTransformResponse

          if (!results) {
            return [response.body]
          }

          return results
        },
        shouldContinue: (item, allItems) => {
          const hasNoPagination = allItems.length > 0 && allItems.every((entry) => entry.id === item.id)
          return !hasNoPagination
        },
        paginate: (response) => {
          const prevSearchParams = response.request.options.searchParams
          const prevPage = Number(prevSearchParams.get(`page`))
          const { total_pages: totalPages } = response.body as Response.PaginatedResponse

          if (prevPage > totalPages) {
            return false
          }

          itemTimer.setStatus(`Getting page ${prevPage}`)

          return {
            searchParams: {
              ...prevSearchParams,
              page: prevPage + 1,
            },
          }
        },
      },
    })

  const fetchData = async ({ url, context }): Promise<Response.ResponseItem> => tmdbGot(url, { context }).json()

  try {
    items = await fetchPaginatedData()

    // The endpoint can also have an "extension" to get additional information from an initial request
    // e.g. getting /account/{account_id}/favorite/tv each TV item only has limited information
    // An additional request for each item against the /tv/:tv_id endpoint is necessary to get all details
    if (endpoint.extension) {
      try {
        const param = getParam(endpoint.extension.url)

        if (!param) {
          gatsbyApi.reporter.panicOnBuild({
            id: ERROR_CODES.getParamUndefined,
            context: {
              sourceMessage: `Couldn't find a parameter for ${endpoint.extension.url} - make sure it uses the format :your_parameter`,
            },
          })
          itemTimer.end()
        }

        const detailedItems = items.map((item) =>
          fetchData({ url: endpoint.extension.url, context: { [param]: item.id } })
        )

        // This replaces the fetched items with the full version
        items = await Promise.all(detailedItems)
      } catch (error) {
        gatsbyApi.reporter.panicOnBuild(
          {
            id: ERROR_CODES.extensionSourcing,
            context: {
              sourceMessage: `Error during sourcing the detailed information (extension) from ${endpoint.url}`,
            },
          },
          error
        )
        itemTimer.end()
      }
    }
  } catch (error) {
    gatsbyApi.reporter.panicOnBuild(
      {
        id: ERROR_CODES.individualSourcing,
        context: {
          sourceMessage: `Error during the sourcing from the API of ${endpoint.url}`,
        },
      },
      error
    )
    itemTimer.end()
  }

  itemTimer.setStatus(`Processing ${items.length} results`)

  items.forEach((item) => {
    const transformedItem = imageTransformation({ node: item, configuration })
    const node = Node({ ...transformedItem, id: item.id.toString() })
    gatsbyApi.actions.createNode(node)
  })

  itemTimer.end()
}
