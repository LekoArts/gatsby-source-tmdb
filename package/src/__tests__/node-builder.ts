import { imageTransformation } from "../node-builder"

const configuration = {
  images: {
    secure_base_url: `https://www.example.com/`,
    backdrop_sizes: [`w100`, `w200`],
    logo_sizes: [`w100`, `w200`],
    poster_sizes: [`w100`, `w200`],
    profile_sizes: [`w100`, `w200`],
    still_sizes: [`w100`, `w200`],
  },
  change_keys: [`foobar`],
}

const node = {
  id: 1,
  backdrop_path: `/backdrop_image.png`,
  logo_path: `/logo_image.png`,
  poster_path: `/poster_image.png`,
  profile_path: `/profile_image.png`,
  still_path: `/still_image.png`,
}

const deepNode = {
  id: 1,
  items: [
    {
      id: 2,
      backdrop_path: `/backdrop_image.png`,
      logo_path: `/logo_image.png`,
    },
    {
      id: 3,
      poster_path: `/poster_image.png`,
      profile_path: `/profile_image.png`,
      still_path: `/still_image.png`,
    },
  ],
}

describe(`imageTransformation in nodeBuilder`, () => {
  it(`should convert nodes one level deep`, () => {
    expect(imageTransformation({ node, configuration })).toStrictEqual({
      id: 1,
      backdrop_path: {
        source: `/backdrop_image.png`,
        w100: `https://www.example.com/w100/backdrop_image.png`,
        w200: `https://www.example.com/w200/backdrop_image.png`,
      },
      logo_path: {
        source: `/logo_image.png`,
        w100: `https://www.example.com/w100/logo_image.png`,
        w200: `https://www.example.com/w200/logo_image.png`,
      },
      poster_path: {
        source: `/poster_image.png`,
        w100: `https://www.example.com/w100/poster_image.png`,
        w200: `https://www.example.com/w200/poster_image.png`,
      },
      profile_path: {
        source: `/profile_image.png`,
        w100: `https://www.example.com/w100/profile_image.png`,
        w200: `https://www.example.com/w200/profile_image.png`,
      },
      still_path: {
        source: `/still_image.png`,
        w100: `https://www.example.com/w100/still_image.png`,
        w200: `https://www.example.com/w200/still_image.png`,
      },
    })
  })
  it(`should convert nodes two levels deep (in items like in lists)`, () => {
    expect(imageTransformation({ node: deepNode, configuration })).toStrictEqual({
      id: 1,
      items: [
        {
          id: 2,
          backdrop_path: {
            source: `/backdrop_image.png`,
            w100: `https://www.example.com/w100/backdrop_image.png`,
            w200: `https://www.example.com/w200/backdrop_image.png`,
          },
          logo_path: {
            source: `/logo_image.png`,
            w100: `https://www.example.com/w100/logo_image.png`,
            w200: `https://www.example.com/w200/logo_image.png`,
          },
        },
        {
          id: 3,
          poster_path: {
            source: `/poster_image.png`,
            w100: `https://www.example.com/w100/poster_image.png`,
            w200: `https://www.example.com/w200/poster_image.png`,
          },
          profile_path: {
            source: `/profile_image.png`,
            w100: `https://www.example.com/w100/profile_image.png`,
            w200: `https://www.example.com/w200/profile_image.png`,
          },
          still_path: {
            source: `/still_image.png`,
            w100: `https://www.example.com/w100/still_image.png`,
            w200: `https://www.example.com/w200/still_image.png`,
          },
        },
      ],
    })
  })
})
