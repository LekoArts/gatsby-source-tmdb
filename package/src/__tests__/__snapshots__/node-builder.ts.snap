// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`imageTransformation in nodeBuilder should convert nodes one level deep 1`] = `
Object {
  "backdrop_path": Object {
    "original": "https://image.tmdb.org/t/p/original/backdrop_image.png",
    "source": "/backdrop_image.png",
    "w1280": "https://image.tmdb.org/t/p/w1280/backdrop_image.png",
    "w300": "https://image.tmdb.org/t/p/w300/backdrop_image.png",
    "w780": "https://image.tmdb.org/t/p/w780/backdrop_image.png",
  },
  "id": 1,
  "logo_path": Object {
    "original": "https://image.tmdb.org/t/p/original/logo_image.png",
    "source": "/logo_image.png",
    "w154": "https://image.tmdb.org/t/p/w154/logo_image.png",
    "w185": "https://image.tmdb.org/t/p/w185/logo_image.png",
    "w300": "https://image.tmdb.org/t/p/w300/logo_image.png",
    "w45": "https://image.tmdb.org/t/p/w45/logo_image.png",
    "w500": "https://image.tmdb.org/t/p/w500/logo_image.png",
    "w92": "https://image.tmdb.org/t/p/w92/logo_image.png",
  },
  "poster_path": Object {
    "original": "https://image.tmdb.org/t/p/original/poster_image.png",
    "source": "/poster_image.png",
    "w154": "https://image.tmdb.org/t/p/w154/poster_image.png",
    "w185": "https://image.tmdb.org/t/p/w185/poster_image.png",
    "w342": "https://image.tmdb.org/t/p/w342/poster_image.png",
    "w500": "https://image.tmdb.org/t/p/w500/poster_image.png",
    "w780": "https://image.tmdb.org/t/p/w780/poster_image.png",
    "w92": "https://image.tmdb.org/t/p/w92/poster_image.png",
  },
  "profile_path": Object {
    "h632": "https://image.tmdb.org/t/p/h632/profile_image.png",
    "original": "https://image.tmdb.org/t/p/original/profile_image.png",
    "source": "/profile_image.png",
    "w185": "https://image.tmdb.org/t/p/w185/profile_image.png",
    "w45": "https://image.tmdb.org/t/p/w45/profile_image.png",
  },
  "still_path": Object {
    "original": "https://image.tmdb.org/t/p/original/still_image.png",
    "source": "/still_image.png",
    "w185": "https://image.tmdb.org/t/p/w185/still_image.png",
    "w300": "https://image.tmdb.org/t/p/w300/still_image.png",
    "w92": "https://image.tmdb.org/t/p/w92/still_image.png",
  },
  "title": "one-level",
}
`;

exports[`imageTransformation in nodeBuilder should convert nodes two levels deep 1`] = `
Object {
  "id": 3,
  "items": Array [
    Object {
      "backdrop_path": Object {
        "original": "https://image.tmdb.org/t/p/original/backdrop_image.png",
        "source": "/backdrop_image.png",
        "w1280": "https://image.tmdb.org/t/p/w1280/backdrop_image.png",
        "w300": "https://image.tmdb.org/t/p/w300/backdrop_image.png",
        "w780": "https://image.tmdb.org/t/p/w780/backdrop_image.png",
      },
      "logo_path": Object {
        "original": "https://image.tmdb.org/t/p/original/logo_image.png",
        "source": "/logo_image.png",
        "w154": "https://image.tmdb.org/t/p/w154/logo_image.png",
        "w185": "https://image.tmdb.org/t/p/w185/logo_image.png",
        "w300": "https://image.tmdb.org/t/p/w300/logo_image.png",
        "w45": "https://image.tmdb.org/t/p/w45/logo_image.png",
        "w500": "https://image.tmdb.org/t/p/w500/logo_image.png",
        "w92": "https://image.tmdb.org/t/p/w92/logo_image.png",
      },
    },
    Object {
      "poster_path": Object {
        "original": "https://image.tmdb.org/t/p/original/poster_image.png",
        "source": "/poster_image.png",
        "w154": "https://image.tmdb.org/t/p/w154/poster_image.png",
        "w185": "https://image.tmdb.org/t/p/w185/poster_image.png",
        "w342": "https://image.tmdb.org/t/p/w342/poster_image.png",
        "w500": "https://image.tmdb.org/t/p/w500/poster_image.png",
        "w780": "https://image.tmdb.org/t/p/w780/poster_image.png",
        "w92": "https://image.tmdb.org/t/p/w92/poster_image.png",
      },
      "profile_path": Object {
        "h632": "https://image.tmdb.org/t/p/h632/profile_image.png",
        "original": "https://image.tmdb.org/t/p/original/profile_image.png",
        "source": "/profile_image.png",
        "w185": "https://image.tmdb.org/t/p/w185/profile_image.png",
        "w45": "https://image.tmdb.org/t/p/w45/profile_image.png",
      },
      "still_path": Object {
        "original": "https://image.tmdb.org/t/p/original/still_image.png",
        "source": "/still_image.png",
        "w185": "https://image.tmdb.org/t/p/w185/still_image.png",
        "w300": "https://image.tmdb.org/t/p/w300/still_image.png",
        "w92": "https://image.tmdb.org/t/p/w92/still_image.png",
      },
    },
  ],
  "title": "two-level",
}
`;

exports[`imageTransformation in nodeBuilder with downloadImages set in endpoint should convert nodes one level deep 1`] = `
Object {
  "backdrop_path": Object {
    "localFile": "local-file-node-id",
    "original": "https://image.tmdb.org/t/p/original/backdrop_image.png",
    "source": "/backdrop_image.png",
    "w1280": "https://image.tmdb.org/t/p/w1280/backdrop_image.png",
    "w300": "https://image.tmdb.org/t/p/w300/backdrop_image.png",
    "w780": "https://image.tmdb.org/t/p/w780/backdrop_image.png",
  },
  "id": 2,
  "title": "img-one-level",
}
`;

exports[`imageTransformation in nodeBuilder with downloadImages set in endpoint should convert nodes two levels deep 1`] = `
Object {
  "id": 4,
  "items": Array [
    Object {
      "backdrop_path": Object {
        "localFile": "local-file-node-id",
        "original": "https://image.tmdb.org/t/p/original/backdrop_image.png",
        "source": "/backdrop_image.png",
        "w1280": "https://image.tmdb.org/t/p/w1280/backdrop_image.png",
        "w300": "https://image.tmdb.org/t/p/w300/backdrop_image.png",
        "w780": "https://image.tmdb.org/t/p/w780/backdrop_image.png",
      },
    },
  ],
  "title": "img-two-level",
}
`;

exports[`imageTransformation in nodeBuilder with downloadImages set in pluginOptions should convert nodes one level deep 1`] = `
Object {
  "backdrop_path": Object {
    "localFile": "local-file-node-id",
    "original": "https://image.tmdb.org/t/p/original/backdrop_image.png",
    "source": "/backdrop_image.png",
    "w1280": "https://image.tmdb.org/t/p/w1280/backdrop_image.png",
    "w300": "https://image.tmdb.org/t/p/w300/backdrop_image.png",
    "w780": "https://image.tmdb.org/t/p/w780/backdrop_image.png",
  },
  "id": 2,
  "title": "img-options-one-level",
}
`;

exports[`imageTransformation in nodeBuilder with downloadImages set in pluginOptions should convert nodes two levels deep 1`] = `
Object {
  "id": 4,
  "items": Array [
    Object {
      "backdrop_path": Object {
        "localFile": "local-file-node-id",
        "original": "https://image.tmdb.org/t/p/original/backdrop_image.png",
        "source": "/backdrop_image.png",
        "w1280": "https://image.tmdb.org/t/p/w1280/backdrop_image.png",
        "w300": "https://image.tmdb.org/t/p/w300/backdrop_image.png",
        "w780": "https://image.tmdb.org/t/p/w780/backdrop_image.png",
      },
    },
  ],
  "title": "img-options-two-level",
}
`;
